backend/src/apiserver/server/artifact_server.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	apiv2beta1 "github.com/kubeflow/pipelines/backend/api/v2beta1/go_client"
+	"github.com/kubeflow/pipelines/backend/src/apiserver/common"
+	"github.com/kubeflow/pipelines/backend/src/apiserver/resource"
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+	"github.com/kubeflow/pipelines/backend/src/v2/objectstore"
+	"github.com/kubeflow/pipelines/third_party/ml-metadata/go/ml_metadata"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	getArtifactRequests = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "artifact_server_get_requests",
+		Help: "The total number of GetArtifact requests",
+	})
+	listArtifactRequests = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "artifact_server_list_requests",
+		Help: "The total number of ListArtifacts requests",
+	})
+)
+
+type ArtifactServerOptions struct {
+	CollectMetrics bool `json:"collect_metrics,omitempty"`
+}
+
+type ArtifactServer struct {
+	resourceManager *resource.ResourceManager
+	options         *ArtifactServerOptions
+}
+
+// Value constraints by MLMD:
+// https://github.com/kubeflow/pipelines/blob/38ef986eaa00e8e8e634a17a7837111b6380685a/third_party/ml-metadata/ml_metadata/proto/metadata_store.proto#L873
+const (
+	DefaultListArtifactSize = 20
+	ArtifactSizeMaximum     = 100
+	ContextSizeMaximum      = 100
+	ArtifactSizeMinimum     = 1
+)
+
+// ListArtifacts lists artifacts. Namespace filtering is assessed based on querying mlmd
+// for all contexts within the provided namespace then fetching all artifacts
+// associated with each context.
+// TODO: Add namespace custom property to artifacts to skip context fetching step.
```

Comment:
as mentioned below I'd like to keep the TODO around for dev references, but I've created a JIRA issue for dsp tracking here: https://issues.redhat.com/browse/RHOAIENG-9776 

we can create one for kfp (though likely this will be addressed as part of the kfp pr when it's sent upstream), when we send it upstream.

Reply:
actually I don't think this is correct, verifying